feat: Shipment Doctype

jbienesdev opened the pull request:
Docs: https://github.com/frappe/erpnext_documentation/pull/208
# Feature Request (Shipment)
- Create Shipment from Delivery Note doctype
- Bind a Shipment to a Delivery Note

## DEMO
### Create your Shipment
![core1](https://user-images.githubusercontent.com/17470909/89377453-6c0ff100-d724-11ea-98f2-084d50f8f8a8.gif)

### Creation of Shipment from Delivery Note
![core3](https://user-images.githubusercontent.com/17470909/89377493-82b64800-d724-11ea-9e61-2c3843fbe8c4.gif)

You can also create shipment from a delivery note. Details and addresses will be carried over.

jbienesdev committed:
feat(shipment): Shipment Doctype with Integrations

jbienesdev commented:
Unit tests are to be followed.

creamdory commented:
@nabinhait Hi! This is the shipment  contribution we discuss to you with @KanchanChauhan. Can you review if there are necessary changes? Thanks!

marination commented on erpnext/erpnext_integrations/doctype/letmeship/letmeship.py:
```diff
@@ -0,0 +1,396 @@
+		'Accept': 'application/json',
+		'Access-Control-Allow-Origin': 'string'
+	}
+	payload = {'pickupInfo': {
```
`get_letmeship_available_services` is very large. Split it into smaller functions. 
- Get payload from a separate function. 
- Reuse the same function for `create_letmeship_shipment` (payload structure is almost identical)

marination commented on erpnext/erpnext_integrations/doctype/letmeship/letmeship.py:
```diff
@@ -0,0 +1,396 @@
+				.format(str(exc)),
+			indicator='orange',
+			alert=True
+		)
```
- Along with general exceptions and a generic message it possible to also catch exceptions specific to (highly likely to occur) while hitting a letmeship endpoint, and catching those and showing a more specific message ?
- Add error logging (`frappe.log_error`) on exception handling

marination commented on erpnext/erpnext_integrations/doctype/letmeship/letmeship.py:
```diff
@@ -0,0 +1,396 @@
+	# LetMeShip have limit of 30 characters for Company field
+	enabled = frappe.db.get_single_value('LetMeShip','enabled')
+	api_id = frappe.db.get_single_value('LetMeShip','api_id')
+	api_password = frappe.db.get_single_value('LetMeShip','api_password')
```
```suggestion
	enabled, api_password, api_id = frappe.db.get_value("LetMeShip", "LetMeShip", ["enabled", "api_password", "api_id"])
```

marination commented on erpnext/erpnext_integrations/doctype/letmeship/letmeship.py:
```diff
@@ -0,0 +1,396 @@
+		'Accept': 'application/json',
+		'Access-Control-Allow-Origin': 'string'
+	}
+	payload = {
```
shift to a function to get payload

marination commented on erpnext/erpnext_integrations/doctype/letmeship/letmeship.py:
```diff
@@ -0,0 +1,396 @@
+			indicator='orange',
+			alert=True
+		)
+
```
- Add error logging
- Similar changes on exception handling as mentioned above

marination commented on erpnext/erpnext_integrations/doctype/letmeship/letmeship.py:
```diff
@@ -0,0 +1,396 @@
+		frappe.throw(
+			_('Error occurred while printing Shipment: {0}')
+				.format(shipment_label_response_data['message'])
+		)
```
add this to a  try-except just like everywhere else.

marination commented on erpnext/erpnext_integrations/doctype/sendcloud/sendcloud.py:
```diff
@@ -0,0 +1,171 @@
+	# Retrieve rates at SendCloud from specification stated.
+	enabled = frappe.db.get_single_value('SendCloud', 'enabled')
+	api_key = frappe.db.get_single_value('SendCloud', 'api_key')
+	api_secret = frappe.db.get_single_value('SendCloud', 'api_secret')
```
```suggestion
	enabled, api_key, api_secret = frappe.db.get_value("SendCloud", "SendCloud", ["enabled", "api_key", "api_secret"])
```

marination commented on erpnext/erpnext_integrations/doctype/letmeship/letmeship.py:
```diff
@@ -0,0 +1,396 @@
+				.format(str(exc)),
+			indicator='orange',
+			alert=True
+		)
```
- Add error logging
- Similar changes on exception handling as mentioned above

marination commented on erpnext/erpnext_integrations/doctype/sendcloud/sendcloud.py:
```diff
@@ -0,0 +1,171 @@
+					available_service.service_id = service['id']
+					available_services.append(available_service)
+		return available_services
+	except Exception as exc:
```
- Add error logging
- Similar changes on exception handling as mentioned above

marination commented on erpnext/erpnext_integrations/doctype/sendcloud/sendcloud.py:
```diff
@@ -0,0 +1,171 @@
+	# Create a transaction at SendCloud
+	enabled = frappe.db.get_single_value('SendCloud', 'enabled')
+	api_key = frappe.db.get_single_value('SendCloud', 'api_key')
+	api_secret = frappe.db.get_single_value('SendCloud', 'api_secret')
```
Can be one consolidated call

marination commented on erpnext/erpnext_integrations/doctype/sendcloud/sendcloud.py:
```diff
@@ -0,0 +1,171 @@
+				'shipment_amount': service_info['total_price'],
+				'awb_number': awb_number
+			}
+	except Exception as exc:
```
- Add error logging
- Similar changes on exception handling as mentioned above

marination commented on erpnext/stock/doctype/delivery_note/delivery_note.py:
```diff
@@ -569,6 +569,59 @@ def make_packing_slip(source_name, target_doc=None):
+			pickup_contact_display += '<br>' + user.mobile_no
+		target.pickup_contact = pickup_contact_display
+
+		contact = frappe.db.get_value("Contact", source.contact_person, ['email_id', 'phone', 'mobile_no'], as_dict=1)
```
If a `source.contact_person` is blank because maybe customer and its contact are not linked this happens:
           `if contact.email_id:`
`AttributeError: 'NoneType' object has no attribute 'email_id'`
Handle this by checking if contact has some value

marination commented on erpnext/stock/doctype/delivery_note/delivery_note.py:
```diff
@@ -569,6 +569,59 @@ def make_packing_slip(source_name, target_doc=None):
+		if user.phone:
+			pickup_contact_display += '<br>' + user.phone
+		if user.mobile_no and not user.phone:
+			pickup_contact_display += '<br>' + user.mobile_no
```
- Force feeding session user details for pickup contact display is not the best way to handle this.
- A field called "Pickup Contact Person" can be added above the `pickup_contact` field where session user can be mapped by default, but, user should be able to change it and select another user too. Pickup Contact Person can differ.
- On change ofcourse `pickup_contact` should change with formatted data. (handle on client side)
- So here, just mapping session user to a new field "Pickup Contact Person" is good enough

marination commented on erpnext/stock/doctype/shipment/shipment.js:
```diff
@@ -0,0 +1,772 @@
+frappe.ui.form.on('Shipment', {
+	setup: function(frm) {
+		if (frm.doc.__islocal) {
+			frm.trigger('pickup_type');
```
no trigger function as such exists

marination commented on erpnext/stock/doctype/shipment/shipment.js:
```diff
@@ -0,0 +1,772 @@
+		}
+		else {
+			frm.toggle_display("pickup_contact_name", false);
+		}
```
This can be handled via **Mandatory Depends On** and **Display Depends On** in the DocType specification at the field level. Will save you quite some lines of code

marination commented on erpnext/stock/doctype/shipment/shipment.js:
```diff
@@ -0,0 +1,772 @@
+		}
+		else {
+			frm.set_df_property("pickup_contact_name", "reqd", 1);
+			frm.toggle_display("pickup_contact_name", true);
```
These can again be handled via Mandatory Depends On and Display Depends On in the DocType specification at the field level

marination commented on erpnext/stock/doctype/shipment/shipment.js:
```diff
@@ -0,0 +1,772 @@
+			frm.toggle_display("shipment_delivery_notes", false);
+		}
+		else {
+			frm.toggle_display("shipment_delivery_notes", true);
```
here as well `toggle_display` and `set_df_property`(reqd) can be handled via Mandatory Depends On and Display Depends On in the DocType specification at the field level

marination commented on erpnext/stock/doctype/shipment/shipment.js:
```diff
@@ -0,0 +1,772 @@
+	},
+	pickup_date: function(frm) {
+		if (frm.doc.pickup_date < frappe.datetime.get_today()) {
+			frappe.throw(__("Pickup Date cannot be in the past"));
```
Make this more specific like 'Pickup Date cannot be lesser than / before current date' , something like that

marination commented on erpnext/stock/doctype/shipment/shipment.js:
```diff
@@ -0,0 +1,772 @@
+		frm.set_value("pickup_contact_name", '');
+		frm.set_value("pickup_address", '');
+		frm.set_value("pickup_contact", '');
+		frm.set_value("pickup_contact_email", '');
```
```suggestion
                let fields = ["pickup_address_name", "pickup_contact_name", "pickup_address", "pickup_contact", "pickup_contact_email"];
                for (let field of fields){
			frm.set_value(field,  '');
                }
```

marination commented on erpnext/stock/doctype/shipment/shipment.js:
```diff
@@ -0,0 +1,772 @@
+		frm.set_value("delivery_contact_name", '');
+		frm.set_value("delivery_address", '');
+		frm.set_value("delivery_contact", '');
+		frm.set_value("delivery_contact_email", '');
```
loop over fields and set as blank

marination commented on erpnext/stock/doctype/shipment/shipment.js:
```diff
@@ -0,0 +1,772 @@
+			let row_index = row.idx - 1;
+			if(validate_duplicate(frm, 'shipment_delivery_notes', row.delivery_note, row_index)) {
+				cur_frm.get_field('shipment_delivery_notes').grid.grid_rows[row_index].remove();
+				frappe.throw(__(`You have entered duplicate Delivery Notes. Please rectify and try again.`));
```
If an error is going to be raised, best is to just tell the user that Row # (row.idx) is a duplicate. 
If deletion of row takes place **before** the error is raised the user will wonder what duplication the system is talking about (theres nothing to rectify)

marination commented on erpnext/stock/doctype/shipment/shipment.js:
```diff
@@ -0,0 +1,772 @@
+			return;
+		}
+	});
+	return duplicate;
```
Just a suggestion, javascript's `some()` method can be used here.

marination commented on erpnext/stock/doctype/shipment/shipment.json:
```diff
@@ -0,0 +1,478 @@
+   "fieldname": "shipment_id",
+   "fieldtype": "Read Only",
+   "label": "Shipment ID"
+  },
```
The field types can be Data or whatever else and `read_only` can be 1. Reason being empty fields are also visible if fieldtype is read only, while that is not the case with the workaround I suggested.
Eats up unnecessary space
![Screenshot 2020-09-16 at 3 17 29 PM](https://user-images.githubusercontent.com/25857446/93321306-ac888180-f82f-11ea-9881-d474d4f9b584.png)

marination commented on erpnext/stock/doctype/shipment/shipment.py:
```diff
@@ -0,0 +1,300 @@
+
+	def validate_weight(self):
+		for parcel in self.shipment_parcel:
+			if parcel.weight <= 0:
```
```suggestion
			if flt(parcel.weight) <= 0:
```

marination commented on erpnext/stock/doctype/shipment/shipment.py:
```diff
@@ -0,0 +1,300 @@
+		frappe.db.set_value('Shipment', shipment, 'carrier_service', shipment_info.get('carrier_service'))
+		frappe.db.set_value('Shipment', shipment, 'shipment_id', shipment_info.get('shipment_id'))
+		frappe.db.set_value('Shipment', shipment, 'shipment_amount', shipment_info.get('shipment_amount'))
+		frappe.db.set_value('Shipment', shipment, 'awb_number', shipment_info.get('awb_number'))
```
fields can be contained in an array and can be looped over for set_value

marination commented on erpnext/stock/doctype/shipment/shipment.json:
```diff
@@ -0,0 +1,478 @@
+   "fieldname": "tracking_url",
+   "fieldtype": "Small Text",
+   "label": "Tracking URL",
+   "read_only": 1
```
Currently a Link is set here , which opens a tab on click. Its cumbersome to scroll all the way down to track.
Instead:
- Store the Link in this field and make it hidden
- Add a button on the form (along with Update Tracking, Print Shipment Label) called **Track Status** which on click will do the same as **Click here to Track Shipment**
- Make two button groups **View** and **Tools**.
-  Under **View** add **Track Status** button
- Undr **Tools** add **Update Tracking**,**Print Shipment Label**

marination commented on erpnext/stock/doctype/shipment/shipment_service_selector.html:
```diff
@@ -0,0 +1,70 @@
+							<td style="width:10%">{{ format_currency(data[i].total_price, 'EUR', 2) }}</td>
+							<td style="width:10%">
+								<button id="data-list-{{i}}" onclick='cur_frm.select_row({{JSON.stringify(data[i])}})' type="button" class="btn btn-success">
+									<span class="glyphicon glyphicon-check"></span>
```
the icon needs a change , also a green icon doesnt fit with the overall product design. Maybe a simple check (without box like current icon) icon would be nice. Also the button bg has to be consistent with the product buttons

marination commented on erpnext/stock/doctype/shipment/shipment_service_selector.html:
```diff
@@ -0,0 +1,70 @@
+{% if (data.length) { %}
```
If possible make sure the data has two sub dictionaries called 'preferred' and 'other' so that if preferred is blank Preferred Services Should not have any header underneath it. 
![Screenshot 2020-09-16 at 4 11 15 PM](https://user-images.githubusercontent.com/25857446/93327775-74397100-f838-11ea-88d6-cd6a6106aee8.png)

You can show a 'No Preferred Services Available'.
In this way jinja doesn't have to loop through every data item to check if it is preferred or not.
Currently it will loop over the entirety of **data** twice

marination commented on erpnext/stock/doctype/shipment_delivery_notes/shipment_delivery_notes.json:
```diff
@@ -0,0 +1,41 @@
+ "modified": "2020-07-09 12:55:01.134270",
+ "modified_by": "Administrator",
+ "module": "Stock",
+ "name": "Shipment Delivery Notes",
```
Rename to **Shipment Delivery Note**. Naming convention accomodates singular only

marination commented on erpnext/stock/doctype/shipment_notification_subscriptions/shipment_notification_subscriptions.json:
```diff
@@ -0,0 +1,40 @@
+ "modified": "2020-07-09 12:55:14.217387",
+ "modified_by": "Administrator",
+ "module": "Stock",
+ "name": "Shipment Notification Subscriptions",
```
Rename to **Shipment Notification Subscription**. Singular

marination commented on erpnext/stock/doctype/shipment_status_update_subscriptions/shipment_status_update_subscriptions.json:
```diff
@@ -0,0 +1,40 @@
+ "modified": "2020-07-09 12:55:27.615463",
+ "modified_by": "Administrator",
+ "module": "Stock",
+ "name": "Shipment Status Update Subscriptions",
```
Rename to **Shipment Status Update Subscription**. Singular

marination commented on erpnext/stock/doctype/shipment/api/utils.py:
```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
```
This entire utils.py file can be kept outside along with shipment.js, shipemnt.py, etc.
Conventionally we do not have a `api` folder inside any doctype
So, just like doctype `pricing_rule` has utils in it, this should also just be kept inside the shipment folder.

marination commented on erpnext/stock/doctype/shipment/shipment.json:
```diff
@@ -0,0 +1,478 @@
+   "label": "Pickup Date",
+   "reqd": 1
+  },
+  {
+   "default": "09:00",
+   "fieldname": "pickup_from",
+   "fieldtype": "Select",
+   "label": "Pickup from",
+   "options": "09:00\n09:30\n10:00\n10:30\n11:00\n11:30\n12:00\n12:30\n13:00\n13:30\n14:00\n14:30\n15:00\n15:30\n16:00\n16:30\n17:00\n17:30\n18:00\n18:30\n19:00"
+  },
+  {
+   "default": "17:00",
+   "fieldname": "pickup_to",
+   "fieldtype": "Select",
+   "label": "Pickup to",
+   "options": "09:00\n09:30\n10:00\n10:30\n11:00\n11:30\n12:00\n12:30\n13:00\n13:30\n14:00\n14:30\n15:00\n15:30\n16:00\n16:30\n17:00\n17:30\n18:00\n18:30\n19:00"
```
Hardcoding is the not the best approach here.
The start time could be 8 am as well, but my options start at 9 am.
Change the fieldtype to **Time** and perform computations on that. The user must have flexibility to choose the start and end time

marination commented on erpnext/stock/doctype/shipment_parcel_template/shipment_parcel_template.json:
```diff
@@ -0,0 +1,78 @@
+   "fieldname": "preset_name",
+   "fieldtype": "Data",
+   "in_list_view": 1,
+   "label": "Preset Name",
```
Parcel Template Name would be better. Not everyone will understand the context of Preset

dottenbr commented:
Thank you for your extensive and well structured code review @marination

jbienesdev commented:
Hi @marination. Updated the PR for review. Thanks!

jbienesdev committed:
fix(shipment): apply code review

marination commented:
@jbienesdev @dottenbr @KanchanChauhan 
So we (@nabinhait, @rmehta and me) had some discussion on how to work integrations going ahead, and after [this](https://discuss.erpnext.com/t/v13-breaking-changes-proposal-for-cleaning-up-naming/66369) post we decided to move all our integrations to a custom app.
Since this PR is not merged yet, it's a good opportunity to move it to a custom app.

The **Shipment** Doctype however seems very useful in the core. So here's what we decided to do : 
## Keep Shipment Doctype in core
- Let the users create shipments from delivery notes
- This means that Shipment Parcel and Shipment Parcel template stay as well.
- Shipment status and notification update subscription can stay along with Shipment Delivery Notes.
- In the Shipment Information section, we can make **Service Provider**, **Shipment ID**, **Shipment Amount**, **Status**, **Carrier**, **Carrier Service**, **AWB Number**, **Tracking Status** editable. So if a user who hasn't installed the custom app just wants to track small shipments and keep record of them manually they can. If they want automation, they will have to install the app they need.

## Move Integrations to Custom App:
- Move Packlink, SendCloud, LetMeShip to one custom app
- Add any missing custom fields apart from whats in the core
- Move **Parcel Service Type**, **Parcel Service**, **Parcel Service Type Alias** to custom app and add as custom doctypes
- Move additional custom buttons in js also to custom app(fetch rates, create shipping label, etc.)

You can absolutely go ahead if this viable to you, or you can let us know if not possible, we will separate and clean it up.

dottenbr commented:
Why would this be in an app? It’s supposed to be a monolithic system. Isn’t this against the core principle of ERPNext?

rmehta commented:
We are beginning to think integrations should be out of the monolithic system because they are hard to maintain and also they bring their own libraries / dependencies that create unwanted bloat.

Will start a discussion on this on the forum.

> On 15-Oct-2020, at 5:35 PM, Dominik Ottenbreit <notifications@github.com> wrote:
> 
> 
> Why would this be in an app? It’s supposed to be a monolithic system. Isn’t this against the core principle of ERPNext?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/frappe/erpnext/pull/22914#issuecomment-709250144>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABCGLCSBPEKP6CDMWFXJILSK3QSLANCNFSM4PVDZ2NQ>.
>

dottenbr commented:
Can we merge this into the current core and then whoever of the frappe team who moves all integrations will move this along with all other integrations?

dottenbr commented:
Also a note: when moving stuff into integrations consider cleaning up pip and node dependencies though patch.

barredterra reviewed (changes requested):
Consider to move these methods into the shipment provider class and let them have the same name (`LetMeShip.create_shipment()` instad of `create_letmeship_shipment()`). The structure of all shipment providers should be like this:

```py
class ShipmentProvider(Document):

	def get_services():
		pass

	def create_shipment():
		pass

	def get_label():
		pass

	def get_tracking_data():
		pass
```

barredterra commented on erpnext/erpnext_integrations/doctype/letmeship/letmeship.json:
```diff
@@ -0,0 +1,55 @@
+  },
+  {
+   "fieldname": "api_password",
+   "fieldtype": "Data",
```
API keys should have the field type "Password".

jbienesdev replied:
Thanks for pointing this out @barredterra. As for the request changes above, right now it isn't really that necessary to change it to a class-based refactor. Also, given that there are discussions above to move parts of this PR to other places for a possible refactor. So, I think we'll leave it for now.

marination commented:
https://github.com/frappe/erpnext-shipping custom app

creamdory commented:
Hi @marination , what is the plan for the implementation? This will be transferred to the custom repo you have created?

marination commented:
@creamdory hey so the plan was elaborated on in a comment above. But to reiterate :
**Shipment** Doctype stays in the core. So shipment form DN stays. We will make fields like AWB number, etc, editable, so that if a user has not installed the custom app, they can use the Shipment Doctype to manually track shipments. So all the tables in the form stay as well.

**Integrations only**, move to a custom app.  So all the **LetMeShip, PackLink** and **SendCloud** files/doctypes will go in the custom app, not the core. The custom buttons (fetch shipping rates, print shipping label, tracking) are ported to the custom app. Parcel Service Doctypes moved to the custom app too.

I will do a code cleanup on the custom app, no issues. But in this very PR , some files and functions will have to be deleted. To be precise, the ones that are in the custom app need to be removed from here. I do not have access to do a cleanup here, will either need access or it would be great if your team did it.

jbienesdev committed:
chore: remove packlink, letmeship, and sendcloud files

jbienesdev committed:
fix(shipment): change shipment test and refactor shipment.js

jbienesdev committed:
chore: remove notification details section

jbienesdev committed:
chore: linter issues and sider checks

jbienesdev committed:
fix: travis build error
- Removed shipment notification and subscription files
- Minor changes on shipment field configuration
- Add shipment to desk

marination commented on erpnext/stock/doctype/shipment/test_shipment.py:
```diff
@@ -229,8 +228,8 @@ def create_shipment_item(item_name, company_name):
 	item.item_name = item_name
 	item.item_code = item_name
 	item.item_group = 'All Item Groups'
-	item.opening_stock = 'Nos'
```
item.stock_uom = 'Nos'

marination commented on erpnext/stock/doctype/shipment/test_shipment.py:
```diff
@@ -229,8 +228,8 @@ def create_shipment_item(item_name, company_name):
 	item.item_group = 'All Item Groups'
-	item.opening_stock = 'Nos'
 	item.standard_rate = 50
+	item.qty = 100
```
remove this line

jbienesdev committed:
Merge branch 'develop' into shipment_with_integration
